core/internal/cltest/cltest.go

```diff
@@ -1700,3 +1700,93 @@ func BatchElemMatchesHash(req rpc.BatchElem, hash common.Hash) bool {
 	return req.Method == "eth_getTransactionReceipt" &&
 		len(req.Args) == 1 && req.Args[0] == hash
 }
+
+type SimulateIncomingHeadsArgs struct {
+	StartBlock, EndBlock int64
+	Interval             time.Duration
+	Timeout              time.Duration
+	HeadTrackables       []strpkg.HeadTrackable
+}
+
+func SimulateIncomingHeads(t *testing.T, args SimulateIncomingHeadsArgs) (cleanup func()) {
+	t.Helper()
+
+	if args.Timeout == 0 {
+		args.Timeout = 60 * time.Second
+	}
+	if args.Interval == 0 {
+		args.Interval = 500 * time.Millisecond
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), args.Timeout)
+	defer cancel()
+	chTimeout := time.After(args.Timeout)
+
+	chDone := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-chDone:
+				return
+			case <-chTimeout:
+				return
+			default:
+				for _, ht := range args.HeadTrackables {
+					ht.OnNewLongestChain(ctx, models.Head{Number: args.StartBlock})
+				}
+				if args.EndBlock >= 0 && args.StartBlock == args.EndBlock {
+					return
+				}
+				args.StartBlock++
+				time.Sleep(args.Interval)
+			}
+		}
+	}()
+	var once sync.Once
```

Comment:
Sometimes it's helpful to be able to forcibly `cancel` it after some condition is met in the test, but you also must have to be able to `defer cancel()` in case the test fails before that point.